op-supervisor/supervisor/backend/db/logs/db.go

```diff
@@ -477,6 +490,45 @@ func (db *DB) Rewind(headBlockNum uint64) error {
 	return nil
 }
 
+// NextExecutingMessage returns the next executing message in the log database.
+// it skips over any non-executing messages, and will return an error if encountered.
+// the iterator is modified in the process.
+func (db *DB) NextExecutingMessage(iter Iterator) (types.ExecutingMessage, error) {
+	db.rwLock.RLock()
+	defer db.rwLock.RUnlock()
+	// this for-loop will break:
+	// - when the iterator reaches the end of the log
+	// - when the iterator reaches an executing message
+	// - if an error occurs
+	for {
+		_, _, _, err := iter.NextLog()
+		if err != nil {
+			return types.ExecutingMessage{}, err
+		}
+		// if the log is not an executing message, both exec and err are empty
+		exec, err := iter.ExecMessage()
+		if err != nil {
+			return types.ExecutingMessage{}, fmt.Errorf("failed to get executing message: %w", err)
+		}
+		if exec != (types.ExecutingMessage{}) {
+			return exec, nil
+		}
+	}
+}
+
+func (db *DB) LastCheckpointBehind(entryIdx entrydb.EntryIdx) (Iterator, error) {
+	for attempts := 0; attempts < searchCheckpointFrequency; attempts++ {
+		// attempt to read the index entry as a search checkpoint
+		_, err := db.readSearchCheckpoint(entryIdx)
+		if err == nil {
```

Comment:
Checking if it's an error that says "not a checkpoint", and only then traversing backwards, would be nice. Other errors, like read errors, can be bubbled up immediately, and should not count as a missing search checkpoint.